fuel-relayer/src/pending_events.rs

```diff
@@ -0,0 +1,375 @@
+    /// This is litlle bit hacky but because we relate validator staking with fuel commit block and not on eth block
+    /// we need to be sure that we are taking proper order of those transactions
+    /// Revert are reported as list of reverted logs in order of Block2Log1,Block2Log2,Block1Log1,Block2Log2.
+    /// I checked this with infura endpoint.
+    bundled_removed_eth_events: Vec<(u64, Vec<EthEventLog>)>,
+    /// finalized validator set
+    finalized_validator_set: HashMap<Address, u64>,
+    /// finalized fuel block
+    finalized_da_height: u64,
+}
+
+/// Pending diff between FuelBlocks
+#[derive(Clone, Debug)]
+pub struct PendingDiff {
+    /// eth block number, It represent until when we are taking stakes and token deposits.
+    /// It is always monotonic and check on its limits are check in consensus and in contract.
+    /// Contract needs to check that when feul block is commited that this number is more then
+    /// finality period N.
+    da_height: u64,
+    /// Validator stake deposit and withdrawel.
+    stake_diff: HashMap<Address, i64>,
+    /// erc-20 pending deposit. deposit nonce.
+    assets_deposited: HashMap<Bytes32, (Address, AssetId, Word)>,
+}
+
+impl PendingDiff {
+    pub fn new(da_height: u64) -> Self {
+        Self {
+            da_height,
+            stake_diff: HashMap::new(),
+            assets_deposited: HashMap::new(),
+        }
+    }
+    pub fn stake_diff(&self) -> &HashMap<Address, i64> {
+        &self.stake_diff
+    }
+    pub fn assets_deposited(&self) -> &HashMap<Bytes32, (Address, AssetId, Word)> {
+        &self.assets_deposited
+    }
+}
+
+impl Default for PendingEvents {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PendingEvents {
+    pub fn new() -> Self {
+        Self {
+            pending: VecDeque::new(),
+            bundled_removed_eth_events: Vec::new(),
+            finalized_validator_set: HashMap::new(),
+            finalized_da_height: 0,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.pending.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.pending.is_empty()
+    }
+
+    pub fn pop_back(&mut self) -> Option<PendingDiff> {
+        self.pending.pop_back()
+    }
+
+    pub fn clear(&mut self) {
+        self.pending.clear()
+    }
+
+    pub fn push_back(&mut self, pending: PendingDiff) {
+        self.pending.push_back(pending)
+    }
+
+    /// Bundle all removed events to apply them in same time when all of them are flushed.
+    /// Done
+    fn bundle_removed_events(&mut self, event: EthEventLog, eth_block: u64) {
+        // agregate all removed events before reverting them.
+        // check if we have pending block for removal
+        if let Some((last_eth_block, list)) = self.bundled_removed_eth_events.last_mut() {
+            // check if last pending block is same as log event that we received.
+            if *last_eth_block == eth_block {
+                // just push it
```

Comment:
Maybe the comments are more verbose than needed

Reply:
hah, this was my thinking so that i dont forget that this is only thing needed :D will remove it